File: beams/beams/custom_scripts/asset_movement/asset_movement.py
import frappe
from frappe.utils import flt

def update_issued_quantity(doc, method):
	"""
	Updates Issued Quantity in 'Required Items Detail' of an Equipment Request,
	and updates or inserts corresponding rows in Project's Allocated Item Details,
	based on the Asset Movement.
	"""
	if not doc.assets:
		frappe.throw("No assets found in this Asset Movement.")

	reference_name = doc.reference_name
	if not reference_name:
		return

	required_items = frappe.get_all(
		"Required Items Detail",
		filters={"parent": reference_name},
		fields=["name", "required_item", "issued_quantity", "required_quantity"]
	)

	if not required_items:
		return

	asset_count = {}
	for asset in doc.assets:
		if not asset.asset_name:
			frappe.throw(f"Asset Name not set for Asset {asset.asset}.")
		asset_count[asset.asset_name] = asset_count.get(asset.asset_name, 0) + 1

	for req in required_items:
		item = req.required_item
		if item in asset_count:
			new_issued_qty = (req.issued_quantity or 0) + asset_count[item]
			frappe.db.set_value("Required Items Detail", req.name, "issued_quantity", new_issued_qty)

	project_name = frappe.db.get_value("Equipment Request", reference_name, "project")
	if not project_name:
		return

	project_doc = frappe.get_doc("Project", project_name)

	for req in required_items:
		item = req.required_item
		if item in asset_count:
			for row in project_doc.allocated_item_details:
				if row.required_item == item:
					row.issued_quantity = (row.issued_quantity or 0) + asset_count[item]
					break
			else:
				project_doc.append("allocated_item_details", {
					"required_item": item,
					"required_quantity": req.required_quantity,
					"issued_quantity": asset_count[item],
				})

	project_doc.save(ignore_permissions=True)

def before_save(doc, method):
	if doc.assets:
		first_asset = doc.assets[0]
		if first_asset.to_employee:
			doc.new_custodian = first_asset.to_employee

			new_custodian_doc = frappe.get_doc("Employee", doc.new_custodian)
			if new_custodian_doc.user_id:
				doc.user_id = new_custodian_doc.user_id

@frappe.whitelist()
def update_asset_location_from_movement(doc, method=None):
	"""
		Updates the location and physical storage details (room, shelf, row, bin)
		of assets listed in an Asset Movement document.
		If any of these fields are empty in the movement record, they will also be cleared in the Asset.
	"""
	if isinstance(doc, str):
		doc = frappe.get_doc("Asset Movement", doc)

	for item in doc.assets:
		if not item.asset:
			continue

		asset = frappe.get_doc("Asset", item.asset)
		updated = False

		if item.department:
			asset.department = item.department
			updated = True

		if item.target_location:
			asset.location = item.target_location
			updated = True
		else:
			return

		for field in ["room", "shelf", "row", "bin"]:
			item_value = getattr(item, field, None)
			setattr(asset, field, item_value if item_value is not None else "")
			updated = True

		if updated:
			asset.save(ignore_permissions=True)


def update_allocated_asset_details(doc, method=None):
	"""Update Asset Request with Asset Movement Item details when Asset Movement is submitted."""

	if doc.reference_doctype != "Asset Request" or not doc.reference_name:
		return

	if doc.purpose == "Issue":

		asset_request = frappe.get_doc("Asset Request", doc.reference_name)

		asset_request.asset_movement_items = []

		for asset_row in doc.assets:
			new_row = asset_request.append("allocated_assets", {})

			new_row.company = asset_row.company
			new_row.asset = asset_row.asset
			new_row.asset_name = asset_row.asset_name
			new_row.source_location = asset_row.source_location
			new_row.target_location = asset_row.target_location
			new_row.from_employee = asset_row.from_employee
			new_row.to_employee = asset_row.to_employee
			new_row.department = asset_row.department
			new_row.room = asset_row.room
			new_row.row = asset_row.row
			new_row.shelf = asset_row.shelf
			new_row.bin = asset_row.bin

		asset_request.save(ignore_permissions=True)



def update_asset_custodian_on_receipt(doc, method=None):

	"""Clears the Asset custodian when an Asset Movement with purpose 'Receipt' is submitted and no new employee is assigned."""

	if doc.purpose == "Receipt" and doc.reference_doctype == "Asset Request":
		for asset_row in doc.assets:
			if asset_row.asset and not asset_row.to_employee:
				frappe.db.set_value("Asset", asset_row.asset, "custodian", None)

File: beams/beams/custom_scripts/asset_movement/asset_movement.js
frappe.ui.form.on('Asset Movement', {
    onload: function(frm) {
        frm.set_query('reference_doctype', function() {
            return { filters: {} };
        });

        frm.set_query('reference_name', function() {
            return { filters: {} };
        });
    },
    refresh: function(frm) {
        frm.trigger('toggle_asset_movement_item_fields');
        // Filter for 'Shelf','Row' & 'Bin' field in assets child table
        frm.fields_dict.assets.grid.get_field('shelf').get_query = function(doc, cdt, cdn) {
            let row = locals[cdt][cdn];
            return {
                filters: {
                    'room': row.room
                }
            };
        };
        frm.fields_dict.assets.grid.get_field('row').get_query = function(doc, cdt, cdn) {
            let row = locals[cdt][cdn];
            return {
                filters: {
                    'shelf': row.shelf
                }
            };
        };
        frm.fields_dict.assets.grid.get_field('bin').get_query = function(doc, cdt, cdn) {
            let row = locals[cdt][cdn];
            return {
                filters: {
                    'row': row.row
                }
            };
        };

        asset_recevied_acknowledgement(frm);
        
    },
    purpose: function(frm) {
        // Call toggle_fields when purpose changes
        frm.trigger('toggle_asset_movement_item_fields');
    },

    toggle_asset_movement_item_fields: function(frm) {
        // Hide Room, Shelf, Row, and Bin fields in Asset Movement Item when purpose is 'Issue'
        const hide_fields = (frm.doc.purpose === 'Issue');

        frm.fields_dict.assets.grid.update_docfield_property('room', 'hidden', hide_fields);
        frm.fields_dict.assets.grid.update_docfield_property('shelf', 'hidden', hide_fields);
        frm.fields_dict.assets.grid.update_docfield_property('row', 'hidden', hide_fields);
        frm.fields_dict.assets.grid.update_docfield_property('bin', 'hidden', hide_fields);
    }
});



function asset_recevied_acknowledgement(frm) {
    frm.clear_custom_buttons();

    const has_unacknowledged = (frm.doc.assets || []).some(row => !row.acknowledged);
    if (frm.doc.docstatus !== 1 || !has_unacknowledged) return;
    if (frappe.session.user !== frm.doc.user_id) return;


    const btn = frm.add_custom_button(__('Acknowledge Receipt'), () => {
        frappe.confirm(
            __('Are you sure you want to acknowledge receipt of these assets?'),
            () => {
                frappe.call({
                    method: "beams.beams.custom_scripts.asset_movement.asset_movement.acknowledge_assets",
                    args: {
                        asset_movement_name: frm.doc.name
                    },
                    callback: function(r) {
                        if (r.message) {
                            frappe.msgprint(r.message);
                            frm.reload_doc();
                        }
                    }
                });
            }
        );
    });

    btn.css({
        backgroundColor: '#2175b1ff',
        color: 'white',
        border: 'none',
        fontWeight: '500'
    });
}



